File: leetcode/cpp/451. 根据字符出现频率排序.cpp
class Solution {
public:
    // 交换函数
    void swap(vector<pair<int, char>>& vec, int a, int b) {
        auto tmp = vec[a];
        vec[a] = vec[b];
        vec[b] = tmp;
    }
    
    // 快排分区函数
    int partition(vector<pair<int, char>>& vec, int l, int h) {
        int start = l - 1;
        for (int i = l; i < h; i++) {
            if (vec[i].first > vec[h].first) {
                swap(vec, i, ++start);
            }
        }
        swap(vec, h, ++start);
        return start;
    }

    // 快排
    void quickSort(vector<pair<int, char>>& vec, int l, int h) {
        if (l < h) {
            int m = partition(vec, l, h);
            quickSort(vec, l, m - 1);
            quickSort(vec, m + 1, h);
        }
    }

    // 根据字符出现频率排序
    string frequencySort(string s) {
        // 特殊情况
        if (s.empty()) {
            return s;
        }

        // 统计hash并且
        unordered_map<char, int> d;
        for (const auto & chr : s) {
            d[chr] += 1;
        }

        // 生成新list
        vector<pair<int, char>> vec;
        for (const auto & it : d) {
            vec.push_back(make_pair(it.second, it.first));
        }

        // 进行排序
        quickSort(vec, 0, vec.size() - 1);

        // 返回结果
        string ret;
        for (const auto & p : vec) {
            string tmp(p.first, p.second);
            ret += tmp;
        }
        return ret;
    }

	// ===== 桶排序 =====
    string frequencySort(string s) {
        // 特殊情况
        if (s.empty()) {
            return s;
        }

        // 统计hash以及出现的最大频次
        int max_cnt = 0;
        unordered_map<char, int> d;
        for (const auto & chr : s) {
            max_cnt = max(max_cnt, ++d[chr]);
        }

        // 进行桶排序
        vector<vector<char>> buckets(max_cnt + 1);
        for (const auto & it : d) {
            buckets[it.second].push_back(it.first);
        }

        // 从后往前合并结果
        string ret;
        for (auto cnt = max_cnt; cnt >= 0; cnt--) {
            for (const auto & chr : buckets[cnt]) {
                string tmp(cnt, chr);
                ret += tmp;
            }
        }
        return ret;
    }
};

File: leetcode/python/451. 根据字符出现频率排序.py
#!/bin/bash python
"""
451. 根据字符出现频率排序

思路:
1. 快排
2. 桶排序
"""

class Solution(object):
    def partition(self, vec, l, h):
        """
        快排分区函数
        """
        start = l
        for i in range(l, h):
            if vec[i][0] > vec[h][0]:
                vec[i], vec[start] = vec[start], vec[i]
                start += 1
        vec[h], vec[start] = vec[start], vec[h]
        return start

    def quickSort(self, vec, l, h):
        """
        快排，从大到小
        vec内元素的个数为：[cnt, char]
        """
        if l < h:
            m = self.partition(vec, l, h)
            self.quickSort(vec, l, m - 1)
            self.quickSort(vec, m + 1, h)

    def frequencySort(self, s):
        """
        :type s: str
        :rtype: str
        先使用快排方法吧
        """
        # 特殊情况
        if s == "":
            return s

        # 统计hash并生成数组, [cnt, char]
        d = dict()
        for char in s:
            d[char] = 1 + d.get(char, 0)
        vec = [(cnt, char) for char, cnt in d.items()]

        # 进行快排
        self.quickSort(vec, 0, len(vec) - 1)

        # 返回结果
        ret = ""
        for cnt, char in vec:
            ret += char * cnt
        return ret

    # ===== 桶排序 =====
    def frequencySort(self, s: str) -> str:
        """
        因为是字符，所以就对128个ASCII进行桶排序即可
        """
        # 特殊情况
        if s == "":
            return s
        
        # 统计hash并且得到最大频次
        max_cnt, d = 0, dict()
        for char in s:
            d[char] = 1 + d.get(char, 0)
            max_cnt = max(max_cnt, d[char])
        
        # 进行桶排序，下标是出现频次，元素是当前字符，可能有多个字符，所以内部还是list
        buckets = [[] for _ in range(max_cnt + 1)]
        for char, cnt in d.items():
            buckets[cnt].append(char)
        
        # 从后往前合成返回结果
        ret = ""
        for cnt in range(max_cnt, -1, -1):
            for char in buckets[cnt]:
                ret += cnt * char
        return ret